### ISSUE TITLE
Extend doesn't seem to work?

### ISSUE THREAD

**Issue by szageda** [stevep99/keyboard-tweaks#9]
I'm trying to implement standard Extend layer for my QWERTY-to-Colemak key remapping, but it doesn't seem to work. Whenever I hold the Extend key (CapsLock), nothing happens. The CapsLock LED doesn't turn on, so AHK appears to be manipulating the CapsLock behaviour, but none of the Extend features work. It still types the regular alphabetical keys.

These are my files:
`main.ahk`
```ahk
; -----------------------------------------------------------------------------
; AUTOHOTKEY SETTINGS
;
; These are generic AutoHotkey software settings, these commands control how
; AHK and the loaded scripts behave while the software is running.

; General
#NoEnv                      ; don't look for environment variables
#KeyHistory 0               ; don't keep key history
#SingleInstance force       ; don't allow scripts to run again
ListLines 0                 ; don't keep executed lines in history

; Performance
Process, Priority,, High    ; sets AHK process priority to high

; Input Mode and Delays
SendMode Input              ; disables all attributes below

; These attributes are kept for troubleshooting and benchmarking purposes,
; but they do not affect how the scripts work if SendMode is set to Input.
SetBatchLines, -1           ; disables 10 ms sleep between line executions
SetDefaultMouseSpeed, 0     ; makes mouse movements instant
SetKeyDelay, -1, -1         ; disables key press delay
SetMouseDelay, -1           ; disables mouse click delay

; Additional Delays
SetControlDelay, -1         ; disables Control command delays
setWinDelay, -1             ; disables window action delays

; -----------------------------------------------------------------------------
; SCRIPTS TO LOAD

; main key mapping
#Include %A_ScriptDir%\keyboard\layouts\qwerty_to_colemak.ahk

; Extend layer
#Include %A_ScriptDir%\keyboard\layers\colemak_extend.ahk

#InputLevel 1
sc03a::F22     ; define Extend on CapsLock
#InputLevel 0
```
`scripts\keyboard\layouts\qwerty_to_colemak.ahk`
```ahk
; Numeric Row
;SC01::`
;SC02::1
;SC03::2
;SC04::3
;SC05::4
;SC06::5
;SC07::6
;SC08::7
;SC09::8
;SC0a::9
;SC0b::0
SC0c::vkDB
SC0d::vkDD

; Top Row
;SC010::q
;SC011::w
SC012::f
SC013::p
SC014::g
SC015::j
SC016::l
SC017::u
SC018::y
SC019::vkDE
SC01a::vkBA
SC01b::vkBB
SC02b::vkDC

; Home Row
;SC01E::a
SC01f::r
SC020::s
SC021::t
SC022::d
;SC023::h
SC024::n
SC025::e
SC026::i
SC027::o
SC028::vkBD

; Bottom Row
;SC02c::z
;SC02d::x
;SC02e::c
;SC02f::v
;SC030::b
SC031::k
;SC032::m
;SC033::,
;SC034::.
;SC035::/
```
`scripts\keyboard\layers\colemak_extend.ahk`
```ahk
; AutoHotKey script for defining an Extend layer (default is CapsLock)
; Hold Extend and press other keys to provide navigation and other fuctionality
;
; The script defines F24 as Extend, so you can apply to any key of your choice
; e.g. to use CapsLock, define:
#InputLevel 1
CapsLock::F24
#InputLevel 0

#Persistent
SetCapsLockState, AlwaysOff

; digit row

F24 & SC001::
return

F24 & SC002::
  Send {Blind}{F1}
return

F24 & SC003::
  Send {Blind}{F2}
return

F24 & SC004::
  Send {Blind}{F3}
return

F24 & SC005::
  Send {Blind}{F4}
return

F24 & SC006::
  Send {Blind}{F5}
return

F24 & SC007::
  Send {Blind}{F6}
return

F24 & SC008::
  Send {Blind}{F7}
return

F24 & SC009::
  Send {Blind}{F8}
return

F24 & SC00A::
  Send {Blind}{F9}
return

F24 & SC00B::
  Send {Blind}{F10}
return

F24 & SC00C::
  Send {Blind}{F11}
return

F24 & SC00D::
  Send {Blind}{F12}
return

; top row

F24 & SC010::
  Send {Esc}
return

F24 & SC011::
  mouseclick, x1
  ;Send {Browser_Back}
return

F24 & SC012::
  Send {CtrlDown}{f}{CtrlUp}
return

F24 & SC013::
  mouseclick, x2
  ;Send {Browser_Forward}
return

F24 & SC014::
  Send {Insert}
return

F24 & SC015::
  Send {Blind}{PgUp}
return

F24 & SC016::
  Send {Blind}{Home}
return

F24 & SC017::
  Send {Blind}{Up}
return

F24 & SC018::
  Send {Blind}{End}
return

F24 & SC019::
  GetKeyState, cp, CapsLock, T
  if cp = D
    SetCapsLockState, AlwaysOff
  else
    SetCapsLockState, AlwaysOn
return

F24 & SC01A::
return

; middle row

F24 & SC01E::
  Send {Blind}{LAltDown}
return

F24 & SC01E Up::
  Send {LAltUp}
return

F24 & SC01F::
  Send {Blind}{LWinDown}
return

F24 & SC01F Up::
  Send {LWinUp}
return

F24 & SC020::
  Send {Blind}{ShiftDown}
return

F24 & SC020 Up::
  Send {ShiftUp}
return

F24 & SC021::
  Send {Blind}{CtrlDown}
return

F24 & SC021 Up::
  Send {CtrlUp}
return

F24 & SC022::
  Send {Blind}{RAltDown}
return

F24 & SC022 Up::
  Send {RAltUp}
return

F24 & SC023::
  Send {Blind}{PgDn}
return

F24 & SC024::
  Send {Blind}{Left}
return

F24 & SC025::
  Send {Blind}{Down}
return

F24 & SC026::
  Send {Blind}{Right}
return

F24 & SC027::
  Send {Blind}{Del}
return

F24 & SC028::
return

; bottom row

F24 & SC056::
  Send {CtrlDown}{y}{CtrlUp}
return

F24 & SC02C::
  Send {CtrlDown}{z}{CtrlUp}
return

F24 & SC02D::
  Send {CtrlDown}{x}{CtrlUp}
return

F24 & SC02E::
  Send {CtrlDown}{c}{CtrlUp}
return

F24 & SC02F::
  Send {CtrlDown}{v}{CtrlUp}
return

F24 & SC030::
return

F24 & SC031::
  Send {Blind}{PrintScreen}
return

F24 & SC032::
  Send {Blind}{Backspace}
return

F24 & SC033::
  Send {Blind}{Tab}
return

F24 & SC034::
  Send {AppsKey}
return

F24 & SC035::
  Send {Return}
return

; RAlt cancel caps / nav layer

RAlt::
  GetKeyState, cp, CapsLock, T
  if navLayer 
  {
    navLayer := 0
  } 
  else if cp = D
  {
    SetCapsLockState, AlwaysOff
  } 
Return
```

Video demo:

https://user-images.githubusercontent.com/84247447/215503078-5d6a37e6-a989-49a9-b21a-3681299c1889.mp4

--
**Specs:**
- Windows 10 x64 21H2
- AutoHotkey 1.1.35.00

**Comment by wobec** [comment]
**UPDATE:**
I've tested the original files without changing anything in them, and they produce the same sympton.

**Comment by stevep99** [comment]
Hi, thanks for the report.
I don't have access to a machine with Windows right now so unfortunately I
can't check this at the moment. But, I noticed you are using F24. I used to
use F24 too but changed it at some point - if I remember correctly, there
was some version of Windows where F24 seemed to cause a conflict.

Also, make sure you have definitions like those in the "config" directory,
which assign the raw keycode to the F20/F22 keys. The intention is to run
the scripts in the config folder which import the Extend/AltGr definitions
that you want to use.

**Comment by wobec** [comment]
Hey,

Yes, I use F24, because that's what is defined in your files. The issue is the same after re-cloning the whole repo again – here's what I've done:
1. Clone the repo.
2. Pointed AutoHotkey to start `config\standard_colemak_config.ahk`.
3. Modified the other AHK scripts location in `#Include`, otherwise I would get this error:
![standard_colemak_config ahk_000289](https://user-images.githubusercontent.com/84247447/217380480-3d3a08fc-1990-4b9b-be1c-cc0b8ba39360.png)
I basically changed `..\keymap\..` to `keyboard-tweaks\keymap\..` in all instances.
4. Tested the script and holding Extend (CapsLock, as mapped in your AHK files) produces the same symptom: It doesn't activate.

By the way, holding AltGr also doesn't produce anything.

Perhaps using CapsLock only and leaving out F22 & F24 keys would fix it? Or perhaps AutoHotkey changed so much during the years that the scripts have become somewhat less compatible, because according to GitHub the files haven't been changed in ~6 years meanwhile AHK has gone through several versions since.

As a side note, AutoHotkey version 2 got released recently, so I might give that a try; although, I'm not sure how it handles version 1.1 code.

**Comment by stevep99** [comment]
I've had another look at this. First, my recollection on the function keys was wrong; it was F22 that had the problem, so the scripts should be using F20 (for AltGr symbols) and F24 (for Extend). There were some incorrect references to F22 still which I have now fixed.

I've tested the updated config scripts on a virtual Windows 10 machine, and they seem to be working for both layers. This was using AutoHotKey v1.1. I downloaded and tried AutoHotKey v2, but that didn't work, so the scripts will need to be migrated to AHK v2 at some point, but for now 1.1 is supported only.
